client/tracing/src/block/mod.rs

```diff
@@ -0,0 +1,132 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Utilities for tracing block execution
+
+use std::sync::Arc;
+
+use parking_lot::Mutex;
+use tracing::{Dispatch, dispatcher};
+use tracing_subscriber::{FmtSubscriber, layer::SubscriberExt, EnvFilter};
+
+use sc_client_api::BlockBackend;
+use sp_api::{Core, Metadata, ProvideRuntimeApi};
+use sp_blockchain::HeaderBackend;
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT, Header},
+};
+use sp_tracing::std_types::Traces;
+
+use crate::{TraceHandler, ProfilingLayer};
+
+// Default to only runtime and state related traces
+const DEFAULT_TARGETS: &'static str = "pallet,frame,state";
+const TRACE_TARGET: &'static str = "block_trace";
+
+/// Holds a reference to the client in order to execute the given block
+/// and record traces for the supplied targets (eg. "pallet,frame,state")
+pub struct BlockExecutor<Block: BlockT, Client> {
+	client: Arc<Client>,
+	block: Block::Hash,
+	targets: Option<String>
+}
+
+impl<Block, Client> BlockExecutor<Block, Client>
+	where
+		Block: BlockT + 'static,
+		Client: HeaderBackend<Block> + BlockBackend<Block> + ProvideRuntimeApi<Block>
+		+ Send + Sync + 'static,
+		Client::Api: Metadata<Block, Error=sp_blockchain::Error>,
+{
+	/// Create a new `BlockExecutor`
+	pub fn new(client: Arc<Client>, block: Block::Hash, targets: Option<String>) -> Self {
+		Self { client, block, targets }
+	}
+
+	/// Execute block, recording all spans and events belonging to `Self::targets`
+	pub fn trace_block(&self) -> Result<Traces, String> {
+		// Prepare block
+		let id = BlockId::Hash(self.block);
+		let extrinsics = self.client.block_body(&id)
+			.map_err(|e| format!("Invalid block id: {:?}", e))?
+			.ok_or("Block not found".to_string())?;
+		let mut header = self.client.header(id)
+			.map_err(|e| format!("Invalid block id: {:?}", e))?
+			.ok_or("Block not found".to_string())?;
+		let parent_hash = header.parent_hash().clone();
+		let parent_id = BlockId::Hash(parent_hash.clone());
+		// Pop digest else RuntimePanic due to: 'Number of digest items must match that calculated.'
+		header.digest_mut().pop();
+		let block = Block::new(header, extrinsics);
+
+		let targets = if let Some(t) = &self.targets { t } else { DEFAULT_TARGETS };
+		let traces = Arc::new(Mutex::new(Traces::default()));
+		let dispatch = create_dispatch(traces.clone(), targets)?;
+
+		if let Err(e) = dispatcher::with_default(&dispatch, || {
+			let span = tracing::info_span!(
+				target: TRACE_TARGET,
+				"trace_block",
+				block_hash = ?self.block,
+				?parent_hash,
+				targets
+			);
+			let _enter = span.enter();
+			self.client.runtime_api().execute_block(&parent_id, block)
+		}) {
+			return Err(format!("Error executing block: {:?}", e));
+		}
+
+		drop(dispatch);
```

Comment:
Is there something I can read to explain how `drop` is being used here? Looking in `tracing` crate docks but don't see a mention of a special `drop` function. I also don't understand where this function is defined or imported.

Reply:
this actually just the `std::mem::drop` function - it's just a way to drop the dispatch early before it goes out of scope so that we can unwrap the `Arc` etc and return an owned `Traces`